### PR TITLE
Inline assembler FPU support

### PR DESCRIPTION
**Pull request by peterhinch**
1. strto_uint_t()  
   Written because I couldn't find a standard library function which could be called without build errors.
2. vmrs()  
   Implemented as a zero argument function for simplicity to copy the FPU status into the ARM status register. I'll provide functionality to copy to other ARM core regsiters if thought worthwhile.
3. functions vcvt_f32_s32 and vcvt_s32_f32  
   Dots replaced with underscores because dots cause problems with interpreter.
   Opcodes correspond with code generated by the C compiler but differ from those in the architecture manual, notably in bit 12 of the top word. I suspect a typo in the manual as the codes in the manual cause a crash.

### CODE REVIEW

**Comment by dpgeorge** [comment]
Looks nice!  I'll review it in detail later.

File: py/emitinlinethumb.c [hunk 1]
```diff
@@ -172,6 +172,46 @@ STATIC const char *get_arg_str(mp_parse_node_t pn) {
...
+        }
+        val = val * 10 + v;
+    }
+    *endptr = (char*)nptr;
+    return val;
+}
+
+STATIC mp_uint_t get_arg_fpreg(emit_inline_asm_t *emit, const char *op, mp_parse_node_t pn) {
+    const char *reg_str = get_arg_str(pn);
+    char *end ;
```
Comment (dpgeorge) [inline py/emitinlinethumb.c]: Need to fix white space before ;.

File: py/emitinlinethumb.c [hunk 2]
```diff
@@ -172,6 +172,46 @@ STATIC const char *get_arg_str(mp_parse_node_t pn) {
...
+    mp_uint_t regno ;
+    regno = 0;
+    if (reg_str[0] == 's'){
```
Comment (dpgeorge) [inline py/emitinlinethumb.c]: Need to add whitespace before {

File: py/emitinlinethumb.c [hunk 3]
```diff
@@ -172,6 +172,46 @@ STATIC const char *get_arg_str(mp_parse_node_t pn) {
...
+        regno =strto_uint_t(reg_str+1, &end) ;
+        if (regno <= 31 && end > reg_str +1)
```
Comment (dpgeorge) [inline py/emitinlinethumb.c]: Needs curly braces.

File: py/emitinlinethumb.c [hunk 4]
```diff
@@ -465,7 +507,8 @@ STATIC void emit_inline_thumb_op(emit_inline_asm_t *emit, qstr op, mp_uint_t n_a
     } else if (n_args == 2) {
         if (MP_PARSE_NODE_IS_ID(pn_args[1])) {
             // second arg is a register (or should be)
-            mp_uint_t op_code, op_code_hi;
+            mp_uint_t op_code, op_code_hi, vd, vm;
```
Comment (dpgeorge) [inline py/emitinlinethumb.c]: vd and vm should be declared below when they are needed.

File: py/emitinlinethumb.c [hunk 5]
```diff
@@ -503,7 +579,7 @@ STATIC void emit_inline_thumb_op(emit_inline_asm_t *emit, qstr op, mp_uint_t n_a
             }
         } else {
             // second arg is not a register
-            mp_uint_t op_code;
+            mp_uint_t op_code, op_code2;
```
Comment (dpgeorge) [inline py/emitinlinethumb.c]: op_code2 is not needed as a variable.

**Comment by dpgeorge** [comment]
vmrs should conform as much as possible to the ARM asm standard.  This is so that when people read the asm code they understand straight away what it does.  So even if you only support certain args, those args still need to be passed.  It would be `vmrs(APSR_nzcv, FPSCR)` and you'd need to verify the args are what you expect.

**Comment by peterhinch** [comment]
Changes requested now implemented. Commit notes provide details.

**Comment by dpgeorge** [comment]
Please remove the `getlbl` pseudo instruction.  That needs to be a separate PR so we can discuss it :)

**Comment by peterhinch** [comment]
Now removed. Learning the ropes, I'm afraid :)

**Comment by dpgeorge** [comment]
Thank you!  I wanted to make this feature a compile-time option (not all MCUs have a VFP unit) so I reworked the code a little and merged in 5008972fefdd0d8cad214d4c0a2fb3daea0ac3c8.

**Pull request closed without merging**